trie.py

```diff
@@ -0,0 +1,100 @@
+
+
+import os
+import sys
+
+
+class TrieNode:
+	"""
+	a node in the trie data structure
+	"""
+	def __init__(self):
+		"""
+		constructor
+		"""
+		self.children = {};
+		self.endNode = False
+
+
+class Trie:
+	"""
+	A trie data structure to store the dictionary
+	"""
+
+	def __init__(self):
+		"""
+		Constructor: create a Trie Root Node.
+		"""
+		self.__root = TrieNode()
+		
+
+	def insert(self, word):
+		"""
+		Insert word into the trie
+		"""
+		if not isinstance(word, str):
+			raise(TypeError, "only string are allowed")
+
+
+		tnode = self.__root
+
+		for letter in word:
+			children = tnode.children
+			if letter not in children.keys():
+				children[letter] = TrieNode();
+			
+			tnode = children[letter];
+
+		tnode.endNode = True;
```

Comment:
i'm wondering about the edge case where an empty string `""` is passed in as the word to insert.  Would this cause the root node of the Trie to incorrectly think its an endNode?  The for loop wouldn't do anything to advance the tnode to point to the children, so it would update the root node, right?

Reply:
@burkstar, you are right. I thought about this. While "" can be considered as a valid word, but do we consider "car   " and "car" as the same word or not ?  My implementation will consider these as two different words. Given this problem I think we will not get into this issue as the puzzle_matrix will not have an "" as an entry in the matrix. 
Looking at just the implementation of the trie, in my opinion considering 
1. "car" 
2. "car   " (word with trailing spaces) or 
3. "   car" 
4. "c a r" ( word when letters are separated with " ") 
5. or any combinations of white spaces and the same letters. 
are all considered as different words, which cannot be called incorrect. I would expect the dictionary to be pre-processed. 
Is the that desirable, may be not . I can pre-process the word to remove the white spaces and then insert into the trie.